llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+//===----- AEIWawRegRewriter.cpp - Rewrite Regs to remove Defs ---------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of a Loop Register rewriting pass, that reduces WAW physical
+// register conflicts, so that software pipelining can operate on loops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseInstrInfo.h"
+#include "AIEBaseRegisterInfo.h"
+#include "Utils/AIELoopUtils.h"
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/CodeGen/LiveDebugVariables.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/LiveRegMatrix.h"
+#include "llvm/CodeGen/LiveStacks.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/MC/MCRegister.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <llvm/CodeGen/MachineBasicBlock.h>
+#include <map>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-waw-reg-rewrite"
+
+namespace {
+
+cl::opt<bool> DisableCopySubRegHandling(
+    "aie-waw-disable-copy-sub-reg-handling",
+    cl::desc("Disable Copy Sub register handling, where WAW regs will not be "
+             "replaced, if reg parts are reused."),
+    cl::init(false), cl::Hidden);
```

Comment:
Nit: I'd prefer to avoid "double negations", maybe rename to e.g. `aie-enable-waw-reg-renaming"

Reply:
I will remove the option fully, since it does not make sense to ever disable it